pkgs/dart_model/lib/src/json.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+
+import 'json_changes.dart';
+
+/// JSON data.
+final class Json {
+  Map<String, Object?> _root;
+
+  Json.fromJson(Map<String, Object?> json) : _root = json;
+
+  JsonChanges computeChangesFrom(Json previous) {
+    final updates = <Update>[];
+    final removals = <Removal>[];
+    _compute(previous._root, _root, Path([]), updates, removals);
+    return JsonChanges(updates: updates, removals: removals);
+  }
+
+  static void _compute(
+      Map<String, Object?> previous,
+      Map<String, Object?> current,
+      Path path,
+      List<Update> updates,
+      List<Removal> removals) {
+    for (final key in previous.keys.followedBy(current.keys).toSet()) {
+      final keyIsInPrevious = previous.containsKey(key);
+      final keyIsInCurrent = current.containsKey(key);
+
+      if (keyIsInPrevious && !keyIsInCurrent) {
+        removals.add(Removal(path: path.followedByOne(key)));
+      } else if (keyIsInPrevious && keyIsInCurrent) {
+        // It's either the same or a change.
+        final previousValue = previous[key]!;
+        final currentValue = current[key]!;
+
+        if (currentValue is Map<String, Object?>) {
+          if (previousValue is Map<String, Object?>) {
+            _compute(previousValue, currentValue, path.followedByOne(key),
+                updates, removals);
+          } else {
+            updates.add(
+                Update(path: path.followedByOne(key), value: currentValue));
+          }
+        } else if (currentValue is String) {
+          if (previousValue is! String || previousValue != currentValue) {
+            updates.add(
+                Update(path: path.followedByOne(key), value: currentValue));
+          }
+        } else if (currentValue is List) {
+          if (previousValue is! List ||
+              !const DeepCollectionEquality()
+                  .equals(previousValue, currentValue)) {
+            updates.add(
+                Update(path: path.followedByOne(key), value: currentValue));
+          }
+        } else {
+          throw UnsupportedError(
+              'Unsupported change: $previousValue to $currentValue');
+        }
+      } else {
+        // It's new.
+        updates.add(Update(path: path.followedByOne(key), value: current[key]));
+      }
+    }
+  }
+
+  /// Returns a new [Json] instance with [changes] made.
+  ///
+  /// TODO(davidmorgan): actually return a new model, copying as needd.
+  Json change(JsonChanges changes) {
+    for (final update in changes.updates) {
+      _updateAtPath(_root, update.path, update.value);
```

Comment:
This model for updates requires re-traversing the model through the same exact paths a lot.

We should structure the updates as a tree, so we can walk the model once and make updates as we go, similar to how the updates themselves are constructed.

Reply:
Ah, yes, that sounds much better, thanks ... let me give that a try.

Reply:
So I tried structuring as maps instead of with paths, indeed it looks better; updated the code.

It does make things more complicated: mostly that it's no longer explicit where the path ends and the update begins, it's implicit in the values.

I'm not sure yet whether it is better to have two trees, one for updates and one for removals; or to merge the trees and use a different value on the leaf node to differentiate. Updates to lists would also need something additional, I guess they are easier to differentiate using a value on the leaf node but that adds significant complexity. For now I think a simple version with maps is a good place to be, added a TODO to answer the next questions :)